File: esp-tracker/supabase/functions/_shared/sms.ts
// SMS transport. Regional choice: for Philippine numbers a local gateway
// (Semaphore, Movider) is dramatically cheaper than Twilio. Verify current
// pricing before committing — this is a recurring cost per alert.

export interface SmsResult { ok: boolean; msgId?: string; error?: string }

const PROVIDER = Deno.env.get("SMS_PROVIDER") ?? "console";

export async function sendSms(to: string, text: string): Promise<SmsResult> {
  try {
    if (PROVIDER === "semaphore") {
      const r = await fetch("https://api.semaphore.co/api/v4/messages", {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: new URLSearchParams({
          apikey: Deno.env.get("SEMAPHORE_API_KEY") ?? "",
          number: to,
          message: text,
          sendername: Deno.env.get("SEMAPHORE_SENDER_NAME") ?? "",
        }),
      });
      if (!r.ok) return { ok: false, error: `semaphore ${r.status}` };
      const body = await r.json();
      return { ok: true, msgId: String(body?.[0]?.message_id ?? "") };
    }

    if (PROVIDER === "twilio") {
      const sid = Deno.env.get("TWILIO_ACCOUNT_SID") ?? "";
      const r = await fetch(
        `https://api.twilio.com/2010-04-01/Accounts/${sid}/Messages.json`, {
        method: "POST",
        headers: {
          Authorization: "Basic " + btoa(`${sid}:${Deno.env.get("TWILIO_AUTH_TOKEN")}`),
          "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams({ To: to, From: Deno.env.get("TWILIO_FROM") ?? "", Body: text }),
      });
      if (!r.ok) return { ok: false, error: `twilio ${r.status}` };
      return { ok: true, msgId: (await r.json()).sid };
    }

    console.log(`[sms:console] -> ${to}: ${text}`);
    return { ok: true, msgId: "console" };
  } catch (e) {
    return { ok: false, error: String(e) };
  }
}

// The downlink. The modem raises RI and wakes even from AT+CSCLK sleep, which
// is what makes "locate now" work without holding a TCP socket open against
// carrier NAT.
export async function sendLocateCommand(msisdn: string): Promise<SmsResult> {
  return sendSms(msisdn, `LOCATE ${Deno.env.get("SMS_CMD_SECRET") ?? ""}`);
}

export function sosText(
  who: string,
  loc: { lat: number; lon: number; accuracy_m: number; source: string } | null,
) {
  return loc
    ? `SOS from ${who}. https://maps.google.com/?q=${loc.lat.toFixed(5)},${loc.lon.toFixed(5)} ` +
      `(+/-${Math.round(loc.accuracy_m)}m, ${loc.source})`
    : `SOS from ${who}. Position not yet known.`;
}

File: esp-tracker/server/app/sms.py
"""SMS transport — used for the escalation ladder and the 'locate now' downlink.

Provider is abstracted because the right choice is regional: for Philippine
numbers a local gateway (Semaphore, Movider) is dramatically cheaper than
Twilio, while Twilio wins on international reach and voice. Verify current
pricing before committing — this is a recurring cost per alert.
"""
import httpx
from .config import (SMS_PROVIDER, SEMAPHORE_API_KEY, SEMAPHORE_SENDER_NAME,
                     TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_FROM)

class SmsResult:
    def __init__(self, ok: bool, msg_id: str = "", error: str = ""):
        self.ok, self.msg_id, self.error = ok, msg_id, error

async def send(to: str, text: str) -> SmsResult:
    try:
        if SMS_PROVIDER == "semaphore":
            async with httpx.AsyncClient(timeout=15) as c:
                r = await c.post("https://api.semaphore.co/api/v4/messages", data={
                    "apikey": SEMAPHORE_API_KEY, "number": to,
                    "message": text, "sendername": SEMAPHORE_SENDER_NAME})
                r.raise_for_status()
                body = r.json()
                mid = str(body[0].get("message_id", "")) if isinstance(body, list) and body else ""
                return SmsResult(True, mid)

        if SMS_PROVIDER == "twilio":
            async with httpx.AsyncClient(timeout=15) as c:
                r = await c.post(
                    f"https://api.twilio.com/2010-04-01/Accounts/{TWILIO_ACCOUNT_SID}/Messages.json",
                    auth=(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN),
                    data={"To": to, "From": TWILIO_FROM, "Body": text})
                r.raise_for_status()
                return SmsResult(True, r.json().get("sid", ""))

        print(f"[sms:console] -> {to}: {text}", flush=True)
        return SmsResult(True, "console")
    except Exception as e:
        return SmsResult(False, error=str(e))

async def send_locate_command(msisdn: str, secret: str) -> SmsResult:
    """Downlink. The modem raises RI and wakes even from AT+CSCLK sleep, which
    is what makes on-demand location work without holding a TCP socket open
    against carrier NAT."""
    return await send(msisdn, f"LOCATE {secret}")
